### ISSUE TITLE
[no-unnecessary-condition] misfires after yield in generator

### ISSUE THREAD

**Issue by laverdet** [typescript-eslint/typescript-eslint#1818]
```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "warn"
  }
}
```

```ts
class Foo {
	flag = false;
	*run() {
		if (this.flag) return 1;
		yield;
		if (this.flag) return 2;
		return 3;
	}
}
```

**Expected Result**
I wouldn't expect an error here because control is yielded back to anyone after the yield.

**Actual Result**
6:7  warning  Unnecessary conditional, value is always falsy  @typescript-eslint/no-unnecessary-condition

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `^2.25.0` |
| `@typescript-eslint/parser`        | `^2.25.0` |
| `TypeScript`                       | `3.8.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `13.11.0` |
| `npm`                              | `6.13.7` |

**Comment by bradzacher** [comment]
Typescript does not do control-flow analysis on the fact that the yield returns to the caller, and the caller could mutate the class property.

The rule just reads the types provided by typescript, and as far as typescript is concerned, the type of the variable has been narrowed to `false`.

Feel free to raise this as a feature request in the typescript repo.
